command/meta.go

```
@@ -599,11 +604,16 @@ func (m *Meta) outputShadowError(err error, output bool) bool {
 // and `terraform workspace delete`.
 const WorkspaceNameEnvVar = "TF_WORKSPACE"
 
+var invalidWorkspaceNameEnvVar = fmt.Errorf("Invalid workspace name set using %s", WorkspaceNameEnvVar)
+
 // Workspace returns the name of the currently configured workspace, corresponding
 // to the desired named state.
-func (m *Meta) Workspace() string {
-	current, _ := m.WorkspaceOverridden()
-	return current
+func (m *Meta) Workspace() (string, error) {
+	current, overridden := m.WorkspaceOverridden()
+	if overridden && !validWorkspaceName(current) {
```

I'm curious why you're only checking if the workspace name is valid when it's the result of an override? (I presume it's checked elsewhere/before this runs, but if so that might be a helpful comment!)

---

This is an extremely good question, and it highlights that I haven't thought this all the way through yet! 🙏 

My concern here was that because of this bug, users may have found themselves with an invalid workspace name. I wanted to make sure that they had a way out of this situation, so if the invalid workspace is selected via the `.terraform/environment` file, I didn't want to throw an error here.

That said, this should at the minimum have a test. And (assuming this path is even the right one) I think it should probably also result in changes to `terraform workspace select` and `delete` so that users can deal with any invalid workspaces they've already created.

I'll think this through further and come up with some next steps. But if all this sounds like a bad idea please do let me know.

---

Here's what I arrived at:

- 7f51be8 adds a test for this branch, which hopefully explains the intent in code. Having an invalid workspace name currently selected (using the `.terraform/environment` file) shouldn't break everything by returning an error from `Meta.Workspace`.
- 512878a ensures that users can delete workspaces which have an invalid name. This really should only have happened as a result of #24564, but it seems to me that we needn't validate workspace name on delete anyway—it's closing the barn door after the horse has been badly named.

Does this all make sense?

---

> it's closing the barn door after the horse has been badly named.
😂 😂 😂 👍